tests/utils.py

```diff
@@ -0,0 +1,143 @@
+""" A file to hold all test utilities. Note the project
+specific overrides are not being used in the GDC currently.
+
+"""
+
+from collections import defaultdict
+import copy
+import os
+import unittest
+import yaml
+
+from jsonschema import validate
+
+from gdcdictionary import SCHEMA_DIR, GDCDictionary
+
+
+def load_yaml(path):
+    with open(path, 'r') as f:
+        return yaml.safe_load(f)
+
+
+DATA_DIR = os.path.join(os.path.dirname(SCHEMA_DIR), 'examples')
+project1 = load_yaml(os.path.join(SCHEMA_DIR, 'projects/project1.yaml'))
+projects = {'project1': project1}
+
+
+class BaseTest(unittest.TestCase):
+
+    def setUp(self):
+        self.dictionary = GDCDictionary(root_dir=SCHEMA_DIR)
+        self.definitions = load_yaml(os.path.join(SCHEMA_DIR, '_definitions.yaml'))
+
+
+def merge_schemas(a, b, path=None):
+    """Recursively zip schemas together
+
+    """
+    path = path if path is not None else []
+    for key in b:
+        if key in a:
+            if isinstance(a[key], dict) and isinstance(b[key], dict):
+                merge_schemas(a[key], b[key], path + [str(key)])
+            elif a[key] == b[key]:
+                pass
+            else:
+                print("Overriding '{}':\n\t- {}\n\t+ {}".format(
+                    '.'.join(path + [str(key)]), a[key], b[key]))
+                a[key] = b[key]
+        else:
+            print("Adding '{}':\n\t+ {}".format(
+                '.'.join(path + [str(key)]), b[key]))
+            a[key] = b[key]
+    return a
+
+
+def get_project_specific_schema(projects, project, schema, entity_type):
+    """Look up the core schema for its type and override it with any
+    project level overrides
+
+    """
+    root = copy.deepcopy(schema)
+    project_overrides = projects.get(project)
+    if project_overrides:
+        overrides = project_overrides.get(entity_type)
+        if overrides:
+            merge_schemas(root, overrides, [entity_type])
+    return root
+
+
+def validate_entity(entity, schemata, project=None, name=''):
+    """Validate an entity by looking up the core schema for its type and
+    overriding it with any project level overrides
+
+    """
+    local_schema = get_project_specific_schema(
+        projects, project, schemata[entity['type']], entity['type'])
+    result = validate(entity, local_schema)
+    return result
+
+
+def validate_schemata(schemata, metaschema):
+    """ Validate schemata"""
+    print('Validating schemas against metaschema... '),
+    for s in schemata.values():
+        validate(s, metaschema)
+
+        def assert_link_is_also_prop(link):
```

Comment:
suggestion: move def outside the for loop. you can use [`partial`](https://docs.python.org/3/library/functools.html#functools.partial) to define it with the s in context